kv/etcd/etcd.go

```diff
@@ -0,0 +1,292 @@
+						time.Sleep(c.opts.WatchChanResetInterval())
+
+						watchChan = c.watcher.Watch(
+							context.Background(),
+							c.opts.KeyFn()(key),
+							clientv3.WithProgressNotify(),
+							clientv3.WithCreatedNotify(),
+						)
+						c.m.etcdWatchReset.Inc(1)
+					}
+				case <-ticker:
+					c.RLock()
+					numWatches := watchable.NumWatches()
+					c.RUnlock()
+
+					if numWatches != 0 {
+						// there are still watches on this watchable, do nothing
+						continue
+					}
+
+					if cleanedUp := c.tryCleanUp(key); cleanedUp {
+						return
+					}
+				}
+			}
+		}()
+	}
+	c.Unlock()
+	_, w, err := watchable.Watch()
+	return w, err
+}
+
+func (c *client) tryCleanUp(key string) bool {
+	c.Lock()
+	defer c.Unlock()
+	watchable, ok := c.watchables[key]
+	if !ok {
+		// not expect this to happen
+		c.logger.Warnf("unexpected: key %s is already cleaned up", key)
+		return true
+	}
+
+	if watchable.NumWatches() != 0 {
+		// a new watch has subscribed to the watchable, do not clean up
+		return false
+	}
+
+	watchable.Close()
+	delete(c.watchables, key)
+	return true
+}
+
+func (c *client) processNotification(r clientv3.WatchResponse, w kv.ValueWatchable, key string) {
+	err := r.Err()
+	if err != nil {
+		c.logger.Errorf("received error on watch channel: %v", err)
+		c.m.etcdWatchError.Inc(1)
+	}
+
+	// we need retry here because if Get() failed on an watch update,
+	// it has to wait 10 mins to be notified to try again
+	if err = c.retrier.Attempt(func() error {
+		return c.update(w, key)
+	}); err != nil {
+		c.logger.Errorf("received notification for key %s, but failed to get value: %v", key, err)
+	}
+}
+
+func (c *client) update(w kv.ValueWatchable, key string) error {
+	v, err := c.Get(key)
+	if err != nil {
+		return err
+	}
+
+	curValue := w.Get()
+	if curValue != nil && curValue.Version() >= v.Version() {
+		return nil
+	}
+
+	return w.Update(v)
+}
+
+func (c *client) Set(key string, v proto.Message) (int, error) {
+	ctx, cancel := c.context()
+	defer cancel()
+
+	value, err := proto.Marshal(v)
+	if err != nil {
+		return 0, err
+	}
+
+	r, err := c.kv.Put(ctx, c.opts.KeyFn()(key), string(value), clientv3.WithPrevKV())
+	if err != nil {
+		c.m.etcdPutError.Inc(1)
+		return 0, err
+	}
+
+	// if there is no prev kv, means this is the first version of the key
+	if r.PrevKv == nil {
+		return 1, nil
+	}
+
+	return int(r.PrevKv.Version + 1), nil
+}
+
+func (c *client) SetIfNotExists(key string, v proto.Message) (int, error) {
+	ctx, cancel := c.context()
+	defer cancel()
+
+	value, err := proto.Marshal(v)
+	if err != nil {
+		return 0, err
+	}
+
+	key = c.opts.KeyFn()(key)
+	r, err := c.kv.Txn(ctx).
+		If(clientv3.Compare(clientv3.Version(key), "=", 0)).
+		Then(clientv3.OpPut(key, string(value))).
+		Commit()
+	if err != nil {
+		c.m.etcdTnxError.Inc(1)
+		return 0, err
+	}
+	if !r.Succeeded {
+		return 0, kv.ErrAlreadyExists
+	}
+	return 1, nil
+}
+
+func (c *client) CheckAndSet(key string, version int, v proto.Message) (int, error) {
```

Comment:
Looks like this is a generic version of `SetIfNotExists`? If so maybe we should change `SetIfNotExists` to
simply call `CheckAndSet(key, 0, v)` instead?